### PR TITLE
Optionally throw exceptions on invalid cookies decoding

### LINKED ISSUE #6730: Cookie decoder stopped throwing exception on malformed cookie name
We want to switch from 4.0 to 4.1, but with 4.1 cookie decoding changed and on illegal characters cookie decoding procedure doesn't throw exception anymore. The piece of code that was throwing exception was in `io.netty.handler.codec.http.cookie.DefaultCookie#DefaultCookie`.

### Expected behavior
When a cookie contains illegal characters in name/value cookie decoder could, based on some parameter, throw an exception, so the server could handle it.
### Actual behavior
There are two modes in CookieDecoder: LAX or STRICT. LAX allowes illegal characters, STRICT swallows malformed cookie. In both cases there is no feedback.
### Steps to reproduce
Tests in `io.netty.handler.codec.http.cookie.ServerCookieDecoderTest` are enough to highlight this situation, specially `io.netty.handler.codec.http.cookie.ServerCookieDecoderTest#testRejectCookieValueWithSemicolon`
### Netty version
Netty 4.1.11.Final

### PR DESCRIPTION
**Pull request by alobodzki**
Motivation:

To provide quick feedback to developers netty based server could fail
requests fast on malformed cookies.

Modifications:

Additional flag on a decoder that would turn on/off throwing exceptions
on cookie validation error.

Result:

Fixes #6730
Developers will have choice whether they want to have a validation that
is swallowing malformed cookies, or get an exception to let the server
handle it.

### CODE REVIEW

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 1]
```diff
@@ -32,27 +32,49 @@
...
 
     protected CookieDecoder(boolean strict) {
-        this.strict = strict;
+        this(strict, false);
     }
 
     protected DefaultCookie initCookie(String header, int nameBegin, int nameEnd, int valueBegin, int valueEnd) {
         if (nameBegin == -1 || nameBegin == nameEnd) {
-            logger.debug("Skipping cookie with null name");
+            if (logger.isDebugEnabled()) {
```
Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: The check for `isDebugEnabled()` is redundant in the such cases. I think, it's should be logging or exception thrown (not both):
```java
private final boolean needReport;

protected CookieDecoder(boolean strict, boolean reportException) {
    this.strict = strict;
    this.reportException = reportException;
    needReport = reportException || logger.isDebugEnabled();
}

private final void reportError(String message) {
    if (reportException) {
        throw new IllegalArgumentException(message);
    } else {
        logger.debug(message);
    }
}

...
if (needReport) {
    reportError("Some error " + xxx + " ...");
}
```

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 2]
```diff
@@ -62,7 +84,12 @@ protected DefaultCookie initCookie(String header, int nameBegin, int nameEnd, in
         if (strict && (invalidOctetPos = firstInvalidCookieNameOctet(name)) >= 0) {
             if (logger.isDebugEnabled()) {
                 logger.debug("Skipping cookie because name '{}' contains invalid char '{}'",
-                        name, name.charAt(invalidOctetPos));
+                        name, invalidOctetPos);
```
Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: For why this change? An `invalidOctetPos` is not a char. The same below.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java [hunk 3]
```diff
@@ -152,4 +152,23 @@ private ServerCookieDecoder(boolean strict) {
 
         return cookies;
     }
+
+    public static class ServerCookieDecoderBuilder {
```
Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java]: final?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java [hunk 4]
```diff
@@ -152,4 +152,23 @@ private ServerCookieDecoder(boolean strict) {
...
+            return this;
+        }
+
+        public ServerCookieDecoderBuilder reportError(boolean reportError) {
+            this.reportError = reportError;
+            return this;
+        }
+
+        public ServerCookieDecoder build() {
+            return new ServerCookieDecoder(strict, reportError);
```
Comment (netkins) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java]: MAJOR Avoid instantiation through private constructors from outside of the constructors class. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AAccessorClassGeneration)

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 5]
```diff
@@ -32,27 +32,49 @@
...
     }
 
     protected DefaultCookie initCookie(String header, int nameBegin, int nameEnd, int valueBegin, int valueEnd) {
         if (nameBegin == -1 || nameBegin == nameEnd) {
-            logger.debug("Skipping cookie with null name");
+            if (logger.isDebugEnabled()) {
+                logger.debug("Skipping cookie with null name");
+            }
+            if (reportException) {
+                throw new IllegalArgumentException("Skipping cookie with null name");
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: Invalid error message: you're not skipping, you're crashing.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 6]
```diff
@@ -32,27 +32,49 @@
...
         }
 
         if (valueBegin == -1) {
-            logger.debug("Skipping cookie with null value");
+            if (logger.isDebugEnabled()) {
+                logger.debug("Skipping cookie with null value");
+            }
+            if (reportException) {
+                throw new IllegalArgumentException("Skipping cookie with null value");
+            }
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: Same as above: redundant check + invalid exception message.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 7]
```diff
@@ -32,27 +32,49 @@
...
-            logger.debug("Skipping cookie because starting quotes are not properly balanced in '{}'",
-                    wrappedValue);
+            if (logger.isDebugEnabled()) {
+                logger.debug("Skipping cookie because starting quotes are not properly balanced in '{}'", wrappedValue);
+            }
+            if (reportException) {
+                throw new IllegalArgumentException(
+                        String.format("Skipping cookie because starting quotes are not properly balanced in '%s'",
+                                wrappedValue));
+            }
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: same as above

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 8]
```diff
@@ -62,7 +84,12 @@ protected DefaultCookie initCookie(String header, int nameBegin, int nameEnd, in
...
             if (logger.isDebugEnabled()) {
                 logger.debug("Skipping cookie because name '{}' contains invalid char '{}'",
-                        name, name.charAt(invalidOctetPos));
+                        name, invalidOctetPos);
+            }
+            if (reportException) {
+                throw new IllegalArgumentException(
+                        String.format("Skipping cookie because name '%s' contains invalid char '%c'",
+                                name, invalidOctetPos));
             }
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: same as above

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 9]
```diff
@@ -72,7 +99,12 @@ protected DefaultCookie initCookie(String header, int nameBegin, int nameEnd, in
         if (strict && (invalidOctetPos = firstInvalidCookieValueOctet(unwrappedValue)) >= 0) {
             if (logger.isDebugEnabled()) {
                 logger.debug("Skipping cookie because value '{}' contains invalid char '{}'",
-                        unwrappedValue, unwrappedValue.charAt(invalidOctetPos));
+                    unwrappedValue, invalidOctetPos);
+            }
+            if (reportException) {
+                throw new IllegalArgumentException(
+                        String.format("Skipping cookie because value '%s' contains invalid char '%c'",
+                                unwrappedValue, invalidOctetPos));
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: same as above

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java (hunk 1 above)
Comment (alobodzki) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: Looks good, but your proposed version introduces computation that is not present in my, or original version as the string will have to be computed, so it will be executed on each request. In my code the JIT will just remove the whole branch and the change will be neutral, do you agree?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java (hunk 2 above)
Comment (alobodzki) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: Good point, I missed that.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java (hunk 3 above)
Comment (alobodzki) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java]: Yes.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java (hunk 1 above)
Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: No.. In my example string computation will be only if `needReport` is true.

Comment (alobodzki) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: Yes, your code is better, cheers.

**Comment by normanmaurer** [comment]
@slandelle happy with it now ?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 10]
```diff
@@ -32,47 +32,61 @@
...
     protected CookieDecoder(boolean strict) {
-        this.strict = strict;
+        this(strict, false);
     }
 
     protected DefaultCookie initCookie(String header, int nameBegin, int nameEnd, int valueBegin, int valueEnd) {
         if (nameBegin == -1 || nameBegin == nameEnd) {
-            logger.debug("Skipping cookie with null name");
+            if (needReport) {
+                reportError(String.format("Invalid cookie with null name found in header '%s'", header));
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: would it make more sense to format these as a `String format, Object... args`? That way you can avoid allocation if the logging level isn't "high" enough.

Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: +1

Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: @Scottmitch logging level is taken into account the `needReport` field :)

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: I would prefer not to have coupling between `needReport` and the logging level. The log methods (`log.debug`, etc..) should be responsible for enforcing whether logging action is taken.

Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: @Scottmitch have you any ideas how avoid duplicate a message computation for logging and exception? It can be simple with lambdas.. But, we can't use lambdas.

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: can you show me how you would do it with lambdas?

Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: @Scottmitch something like this:
```java
...
reportError(() -> String.format("Invalid cookie with null name found in header '%s'", header));
...

private void reportError(Supplier<String> messageSupplier) {
    if (reportException) {
        throw new IllegalArgumentException(messageSupplier.get());
    }
    if (logger.isDebugEnabled()) {
        logger.debug(messageSupplier.get());
    }
}
```

Such approach (lazy string computation through lambdas) is used e.g. in the `log4j2` [1].

[1] https://logging.apache.org/log4j/2.x/manual/api.html#LambdaSupport

Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: I propose to replace the field `needReport` with the method:
```java
private boolean isNeedReportError() {
    return reportException || logger.isDebugEnabled();
}
```
This will fix the problem with changing logging level in runtime.

Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: There is another option with override..

in the `CookieDecoder`:
```java
protected boolean isNeedReportError() {
    return logger.isDebugEnabled();
}

protected void reportError(String message) {
    logger.debug(message);
}

...
if (isNeedReportError()) {
    reportError(String.format("..."));
}
...
```
in the `ServerCookieDecoder`:
```java
public static final ServerCookieDecoder FAIL_FAST = new ServerCookieDecoder(true) {
    @Override
    protected boolean isNeedReportError() {
        return true;
    }

    @Override
    protected void reportError(String message) {
        throw new IllegalArgumentException(message);
    }
}
```


I like this more.

Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: @Scottmitch WDYT ?

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: @fenik17 - sounds reasonable. Lambdas in this case are mostly syntactic sugar. You can just create your own interface and implement the necessary methods. We have our own `Supplier`s (e.g. `BooleanSupplier`) and intend to remove them in favor of the JDK8's when we can drop support for older JDK versions ... you can create an internal interface or maybe even consider a more visible `Supplier<T>` if necessary.

Comment (fenik17) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: @Scottmitch I can provide PR with the such approach after the current PR will be merged.

Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: followup PR sgtm

**Comment by alobodzki** [comment]
@fenik17 I like the lambda approach better, doesn't make sense to merge this commit and build on top of it. I added some example implementation of this feature.

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java [hunk 11]
```diff
@@ -152,4 +156,23 @@ private ServerCookieDecoder(boolean strict) {
...
+            return this;
+        }
+
+        public ServerCookieDecoderBuilder cookieErrorHandler(CookieErrorHandler cookieErrorHandler) {
+            this.cookieErrorHandler = cookieErrorHandler;
+            return this;
+        }
+
+        public ServerCookieDecoder build() {
+            return new ServerCookieDecoder(this);
```
Comment (netkins) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/ServerCookieDecoder.java]: MAJOR Avoid instantiation through private constructors from outside of the constructors class. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AAccessorClassGeneration)
MAJOR Avoid instantiation through private constructors from outside of the constructors class. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=pmd%3AAccessorClassGeneration)

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieDecoderTest.java [hunk 12]
```diff
@@ -99,49 +101,49 @@ public void testDecodingGoogleAnalyticsCookie() {
     public void testDecodingLongValue() {
         String longValue =
                 "b___$Q__$ha__<NC=MN(F__%#4__<NC=MN(F__2_d____#=IvZB__2_F____'=KqtH__2-9____" +
-                "'=IvZM__3f:____$=HbQW__3g'____%=J^wI__3g-____%=J^wI__3g1____$=HbQW__3g2____" +
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieDecoderTest.java]: please revert formatting changes

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieDecoderTest.java [hunk 13]
```diff
@@ -56,12 +57,12 @@ public void testDecodingMultipleCookies() {
...
-            "__utma=48461872.1094088325.1258140131.1258140131.1258140131.1; " +
-            "__utmb=48461872.13.10.1258140131; __utmc=48461872; " +
-            "__utmz=48461872.1258140131.1.1.utmcsr=overstock.com|utmccn=(referral)|" +
-                    "utmcmd=referral|utmcct=/Home-Garden/Furniture/Clearance/clearance/32/dept.html";
+                "ARPT=LWUKQPSWRTUN04CKKJI; " +
+                        "kw-2E343B92-B097-442c-BFA5-BE371E0325A2=unfinished_furniture; " +
+                        "__utma=48461872.1094088325.1258140131.1258140131.1258140131.1; " +
+                        "__utmb=48461872.13.10.1258140131; __utmc=48461872; " +
+                        "__utmz=48461872.1258140131.1.1.utmcsr=overstock.com|utmccn=(referral)|" +
+                        "utmcmd=referral|utmcct=/Home-Garden/Furniture/Clearance/clearance/32/dept.html";
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieDecoderTest.java]: please revert formatting changes

File: codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieDecoderTest.java [hunk 14]
```diff
@@ -173,10 +175,36 @@ public void testDecodingOldRFC2965Cookies() {
 
     @Test
     public void testRejectCookieValueWithSemicolon() {
-        Set<Cookie> cookies = ServerCookieDecoder.STRICT.decode("name=\"foo;bar\";");
+        Set<Cookie> cookies = ServerCookieDecoder.STRICT.decode("name=\"foo,bar\";");
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/cookie/ServerCookieDecoderTest.java]: revert ?

File: codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java [hunk 15]
```diff
@@ -81,4 +91,8 @@ protected DefaultCookie initCookie(String header, int nameBegin, int nameEnd, in
         cookie.setWrap(wrap);
         return cookie;
     }
+
+    public interface CookieErrorHandler {
```
Comment (Scottmitch) [inline codec-http/src/main/java/io/netty/handler/codec/http/cookie/CookieDecoder.java]: As discussed in https://github.com/netty/netty/pull/6744/#discussion_r116452809 I think adding a `STRICT` and `LAX` options should be good enough and we can make this interface an implementation detail (not expose it publicly).

**Comment by normanmaurer** [comment]
@alobodzki ping

**Comment by normanmaurer** [comment]
Closing this due no activity.
